pkg/migrate/apisimigration.go

```diff
@@ -0,0 +1,184 @@
+	migration
+	logger log.FieldLogger
+}
+
+// NewAPISIMigration creates a new APISIMigration
+func NewAPISIMigration(client client, cfg config.CentralConfig) *APISIMigration {
+	logger := log.NewFieldLogger().
+		WithPackage("sdk.migrate").
+		WithComponent("instance-migration")
+
+	return &APISIMigration{
+		migration: migration{
+			client: client,
+			cfg:    cfg,
+		},
+		logger: logger,
+	}
+}
+
+// Migrate checks an APIServiceInstance for the "scopes" key in the schema, and removes it if it is found.
+func (m *APISIMigration) Migrate(ri *v1.ResourceInstance) (*v1.ResourceInstance, error) {
+	if ri.Kind != mv1a.APIServiceGVK().Kind {
+		return ri, nil
+	}
+
+	// skip migration if instance migration is not enabled
+	if !m.cfg.GetMigrationSettings().ShouldCleanInstances() {
+		return ri, nil
+	}
+
+	logger := m.logger.WithField(serviceName, ri.Name)
+
+	// skip migration if instance migration has been completed
+	details := util.GetAgentDetails(ri)
+	if len(details) > 0 {
+		completed := details[definitions.InstanceMigration]
+		if completed == definitions.MigrationCompleted {
+			// migration ran already
+			logger.Debugf("service instance migration already completed")
+			return ri, nil
+		}
+	}
+
+	// get all revisions for this service
+	revisions, err := m.getRevisions(ri)
+	if err != nil {
+		return ri, err
+	}
+	logger.WithField("revisions", revisions).Debug("all revisions")
+
+	// get all instances for each revision
+	wg := &sync.WaitGroup{}
+	errCh := make(chan error, len(revisions))
+	instances := []*v1.ResourceInstance{}
+	instancesLock := sync.RWMutex{}
+
+	for _, rev := range revisions {
+		wg.Add(1)
+
+		go func(r *v1.ResourceInstance) {
+			defer wg.Done()
+
+			revisionInstances, err := m.getInstances(r)
+
+			instancesLock.Lock()
+			defer instancesLock.Unlock()
+			instances = append(instances, revisionInstances...)
+
+			errCh <- err
+		}(rev)
+	}
+
+	wg.Wait()
+	close(errCh)
+
+	for e := range errCh {
+		if e != nil {
+			return ri, e
+		}
+	}
+	logger.WithField("instances", instances).Debug("all instances")
+
+	err = m.cleanInstances(logger, instances)
+	if err != nil {
+		return ri, err
+	}
+
+	// mark the migration as complete
+	util.SetAgentDetailsKey(ri, definitions.InstanceMigration, definitions.MigrationCompleted)
+	err = m.updateRI(ri)
+	return ri, err
+}
+
+// updateRev gets a list of revisions for the service
+func (m *APISIMigration) getRevisions(ri *v1.ResourceInstance) ([]*v1.ResourceInstance, error) {
+	url := m.cfg.GetRevisionsURL()
+	q := map[string]string{
+		"query": queryFunc(ri.Name),
+	}
+
+	return m.getAllRI(url, q)
+}
+
+// updateRev gets a list of revisions for the service
+func (m *APISIMigration) getInstances(ri *v1.ResourceInstance) ([]*v1.ResourceInstance, error) {
+	url := m.cfg.GetInstancesURL()
+	q := map[string]string{
+		"query": queryFunc(ri.Name),
+	}
+
+	return m.getAllRI(url, q)
+}
+
+func (m *APISIMigration) cleanInstances(logger log.FieldLogger, instances []*v1.ResourceInstance) error {
```

Comment:
anything that could be happening to these instances at the time cleanup is happening?

Reply:
right, maybe not discovery.  Don't know if there were any HC's or anything else that could be messing with the same instances.

Reply:
I do not think any HCs touch instances